core/pkg/artifacts/downloader.go

```diff
@@ -102,49 +142,52 @@ func (ad *ArtifactDownloader) downloadFiles(artifactID string, manifest Manifest
 	nameToScheduledTime := map[string]time.Time{}
 	taskResultsChan := make(chan TaskResult)
 	manifestEntriesBatch := make([]ManifestEntry, 0, batchSize)
+	manifestEntriesCopy := map[string]ManifestEntry{}
+	maps.Copy(manifestEntriesCopy, manifestEntries)
 
 	for numDone < len(manifestEntries) {
-		var cursor *string
 		hasNextPage := true
 		for hasNextPage {
 			// Prepare a batch
 			now := time.Now()
 			manifestEntriesBatch = manifestEntriesBatch[:0]
-			response, err := gql.ArtifactFileURLs(
-				ad.Ctx,
-				ad.GraphqlClient,
-				artifactID,
-				cursor,
-				&batchSize,
-			)
-			if err != nil {
-				return err
-			}
-			hasNextPage = response.Artifact.Files.PageInfo.HasNextPage
-			cursor = response.Artifact.Files.PageInfo.EndCursor
-			for _, edge := range response.GetArtifact().GetFiles().Edges {
-				filePath := edge.GetNode().Name
-				entry, err := manifest.GetManifestEntryFromArtifactFilePath(filePath)
-				if err != nil {
-					return err
-				}
+			curBatchSize := 0
+			var entriesToFetch []gql.ArtifactManifestEntryInput
+			for filePath, entry := range manifestEntriesCopy {
 				if _, ok := nameToScheduledTime[filePath]; ok {
 					continue
 				}
-				// Reference artifacts will temporarily be handled by the python user process
 				if entry.Ref != nil {
+					// Reference artifacts will temporarily be handled by the python user process
+					// entry.LocalPath = &filePath
+					// nameToScheduledTime[*entry.LocalPath] = now
+					// manifestEntriesBatch = append(manifestEntriesBatch, entry)
 					numDone++
+					delete(manifestEntriesCopy, filePath)
 					continue
+				} else {
+					entryInput := gql.ArtifactManifestEntryInput{
+						Name:            filePath,
+						Digest:          entry.Digest,
+						BirthArtifactID: entry.BirthArtifactID,
+						Size:            &entry.Size,
+					}
+					entriesToFetch = append(entriesToFetch, entryInput)
+					nameToScheduledTime[*entry.LocalPath] = now
+				}
+				curBatchSize += 1
+				if curBatchSize >= batchSize {
+					break
 				}
-				node := edge.GetNode()
-				if node == nil {
-					return fmt.Errorf("error reading entry from fetched file urls")
+			}
+			if len(entriesToFetch) > 0 {
+				entries, err := ad.getEntriesWithDownloadURLs(artifactID, entriesToFetch, manifest, manifestEntriesCopy)
```

Comment:
Do we need to make sure this is backwards compatible? 

We might have to add some server introspection here so that in case its an old server version,  we can revert to using the existing method of fetching.